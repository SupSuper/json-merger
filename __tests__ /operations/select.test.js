const jsonMerger = require("../../dist");
const {testConfig} = require("../__helpers__");

describe("when merging two objects and a source object contains a $select operation", function () {

    test("and $select.pointer is set it should use the json pointer to select a value in the target", function () {

        const object1 = {
            "a": [
                "should not be selected",
                "should be selected"
            ]
        };

        const object2 = {
            "a": {
                "$select": {
                    "pointer": "/1"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $select.path is set it should use the json path to select a value in the target", function () {

        const object1 = {
            "a": [
                "should not be selected",
                "should be selected"
            ]
        };

        const object2 = {
            "a": {
                "$select": {
                    "path": "$[1]"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and a $select.path matching multiple elements is set but $select.multiple is not set it should return the first value", function () {

        const object1 = {
            "a": [
                "should be selected",
                "should not be selected"
            ]
        };

        const object2 = {
            "a": {
                "$select": {
                    "path": "$[*]"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and a $select.path matching multiple elements is set and $select.multiple is true it should return an array with all matches", function () {

        const object1 = {
            "a": [
                "should be selected",
                "should also be selected"
            ]
        };

        const object2 = {
            "a": {
                "$select": {
                    "path": "$[*]",
                    "multiple": true
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $select.from is set to 'target' it should select from the target", function () {

        const object1 = {
            "a": [
                "target"
            ]
        };

        const object2 = {
            "a": {
                "$select": {
                    "from": "target",
                    "pointer": "/"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $select.from is set to 'targetRoot' it should select from the target root", function () {

        const object1 = {
            "a": [
                "target"
            ]
        };

        const object2 = {
            "a": {
                "$select": {
                    "from": "targetRoot",
                    "pointer": "/"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $select.from is set to 'source' it should select from the source", function () {

        const object1 = {
            "a": [
                "target"
            ]
        };

        const object2 = {
            "a": {
                "$select": {
                    "from": "source",
                    "pointer": "/"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $select.from is set to 'sourceRoot' it should select from the source root", function () {

        const object1 = {
            "a": [
                "target"
            ]
        };

        const object2 = {
            "a": {
                "$select": {
                    "from": "sourceRoot",
                    "pointer": "/"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $select.from is set to a different value it should process the $select.from value and select from the result", function () {

        const object1 = {
            "a": [
                "target"
            ]
        };

        const object2 = {
            "a": {
                "$select": {
                    "from": {
                        "b": {
                            "$replace": "replaced"
                        }
                    },
                    "pointer": "/"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });
});
