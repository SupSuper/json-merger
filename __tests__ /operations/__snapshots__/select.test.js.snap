// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when merging two objects and a source object contains a $select operation and $select.from is set it should process the $select.from value and select from the result 1`] = `
"{
	\\"targetProp\\": \\"object1.targetProp\\",
	\\"sharedProp\\": \\"should be the value of sharedProp\\"
}"
`;

exports[`when merging two objects and a source object contains a $select operation and $select.path is set it should use the json pointer to select a value in the target 1`] = `
"{
	\\"a\\": \\"should be selected\\"
}"
`;

exports[`when merging two objects and a source object contains a $select operation and $select.query is set it should use the json path to select a value in the target 1`] = `
"{
	\\"a\\": \\"should be selected\\"
}"
`;

exports[`when merging two objects and a source object contains a $select operation and a $select.query matching multiple elements is set and $select.multiple is true it should return an array with all matches 1`] = `
"{
	\\"targetProp\\": [
		\\"should be selected\\",
		\\"should also be selected\\"
	],
	\\"sharedProp\\": [
		\\"should be selected\\",
		\\"should also be selected\\",
		\\"should be merged and visible\\"
	]
}"
`;

exports[`when merging two objects and a source object contains a $select operation and a $select.query matching multiple elements is set but $select.multiple is not set it should return the first value 1`] = `
"{
	\\"targetProp\\": [
		\\"should be the value of sourceProp\\",
		\\"object1/targetProp/1\\"
	],
	\\"sourceProp\\": \\"should be the value of sourceProp\\"
}"
`;
